tool/tbot/botfs/fs_linux.go

```diff
@@ -0,0 +1,454 @@
+			if entry.Perms != botAndReaderMode {
+				errors = append(errors, trace.BadParameter("mask entry has improper permissions %d", entry.Perms))
+			}
+		case acl.TagGroup:
+			// Group tags currently not allowed.
+			errors = append(errors, trace.BadParameter("unexpected group entry found"))
+		case acl.TagUser:
+			userTagCount++
+
+			// It's only worth checking the qualifiers if we know all expected
+			// values in advance. We can't know them at bot runtime in any way
+			// that isn't also subject to e.g. an attacker with root / owner
+			// access, so we might as well not spam warnings every time we
+			// verify the ACLs. We'll have to rely on the tag counter instead.
+			if opts.BotUser != nil &&
+				opts.ReaderUser != nil &&
+				entry.Qualifier != opts.BotUser.Uid &&
+				entry.Qualifier != opts.ReaderUser.Uid {
+				errors = append(errors, trace.BadParameter("invalid qualifier %q for user entry", entry.Qualifier))
+			}
+
+			if entry.Perms != botAndReaderMode {
+				errors = append(errors, trace.BadParameter("invalid permissions %q for bot user entry", entry.Perms))
+			}
+		}
+	}
+
+	if userTagCount > maxExpectedUserTags {
+		errors = append(errors, trace.BadParameter("unexpected number of user tags"))
+	}
+
+	return trace.NewAggregate(errors...)
+}
+
+// ConfigureACL configures ACLs of the given file to allow writes from the bot
+// user.
+func ConfigureACL(path string, owner *user.User, opts *ACLOptions) error {
+	if owner.Uid == opts.BotUser.Uid && owner.Uid == opts.ReaderUser.Uid {
+		// We'll end up with an empty ACL. This isn't technically a problem
+		log.Warnf("The owner, bot, and reader all appear to be the same "+
+			"user (%+v). This is an unusual configuration: consider setting "+
+			"`acls: off` in the destination config to remove this warning.",
+			owner.Username)
+	}
+
+	// We fully specify the ACL here to ensure the correct permissions are
+	// always set (rather than just appending an "allow" for the bot user).
+	// Note: These need to be sorted by tag value.
+	ownerMode, botAndReaderMode, err := desiredPerms(path)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Note: ACL entries need to be ordered per acl_linux entryPriority
+	desiredACL := acl.ACL{
+		{
+			// Note: Mask does not apply to user object.
+			Tag:   acl.TagUserObj,
+			Perms: ownerMode,
+		},
+	}
+
+	// Only add an entry for the bot user if it isn't the owner.
+	if owner.Uid != opts.BotUser.Uid {
+		desiredACL = append(desiredACL, acl.Entry{
+			// Entry to allow the bot to read/write.
+			Tag:       acl.TagUser,
+			Qualifier: opts.BotUser.Uid,
+			Perms:     botAndReaderMode,
+		})
+	}
+
+	// Only add entry for the reader if it isn't the owner.
+	if owner.Uid != opts.ReaderUser.Uid {
+		desiredACL = append(desiredACL, acl.Entry{
+			// Entry to allow the reader user to read/write.
+			Tag:       acl.TagUser,
+			Qualifier: opts.ReaderUser.Uid,
+			Perms:     botAndReaderMode,
+		})
+	}
+
+	desiredACL = append(desiredACL, []acl.Entry{
```

Comment:
No need for the slice literal, right?

```
foo = append(foo, []int{1, 2, 3}...)
```

can be simplified to

```
foo = append(foo, 1, 2, 3)
```

Reply:
I thought that ended up being syntactically nicer but it turns out I was just being dumb. Will fix.